### PR TITLE
Eradicate the IdList bottleneck

### LINKED ISSUE #932: Poor performance when extruding translate groups
### System information

- solvespace from master @ 7795b0bc1911a01eaa4661e252488d8f07ec2135
- Debian testing, amd64
- gcc 10.2.1

### Details

I was doing a small project for fun, and went a bit over board with some holes in a sheet metal, to the point where it takes about 4 minutes to render do anything useful.

I used an extrusion of translate of translate to generate a 2D pattern.

[case-vents.slvs.gz](https://github.com/solvespace/solvespace/files/5962704/case-vents.slvs.gz)  (~20MB unpacked, can take few minutes to open, be warned).

Looking at `perf top`, and a bit of code in `Generate`, it looks like the major culprit it CopyEntity, which triggers a copy of Entity classes. The base class (`EntityBase`) has two `std::string` fields `str` and `font`, which I believe should not be in the `EntityBase`, as most of the entities doesn't use these fields actually. Because they are there directly, each construction and destruction, will be called every time any entity is copied (I don't know if it will also trigger malloc/free, which could be even worse). In perf I see `~basic_string<...>`, using about 90% of the CPU. `Platform::Path file` from EntityBase, also shows quite a lot in `perf top`. Other option is to use reference counting for strings maybe. Not sure.

### LINKED ISSUE #939: Poor Linking performance
I have noticed that the more complex the components become the slower the performance gets in solvespace. This is particularly apparent when linking said complex components. Just changing a size constraint can take up to 10 seconds to complete.

Do you know what might be causing this?
Is there anything I can do to get the most performance out of solvespace?

### PR DESCRIPTION
**Pull request by pjanx**
Should fix #939 and fix #932 almost entirely, shifting the bottleneck for large models. Passes tests and memcheck.

Properties of this changeset:
 - memory complexity of IdLists increases by `n * sizeof(int)` bytes, which is about as low as you can go here
 - `IdList::Add` now relies on a single `memmove` of PODs, which is faster than `std::move`-ing singular `elems`
 - `IdList::RemoveTagged` uses a bit of additional temporary memory to keep its complexity

Overall, I'm happy with this now, and I wouldn't be afraid of merging it. Commits should be reasonably atomic.

I think I inadvertently replicated half of [Evil-Spirit's change](https://github.com/Evil-Spirit/solvespace-master/commit/44eb1a4703a1be50d74fdfeacb97f89c06e08ab6). I haven't understood yet what he's doing there, though, even if it looks vaguely similar.

### CODE REVIEW

**Comment by pjanx** [comment]
I would appreciate it if someone tried to resolve the Windows build failure, I don't speak MSVC++ error messages.

**Comment by ruevs** [comment]
:-D everyone comes to more or less the same idea :-) I'll push mine tomorrow - just for comparison... But I haven't finished the iterator.

**Comment by ruevs** [comment]
Here is mine https://github.com/ruevs/solvespace/tree/FixIdList_TheNextAttempt working and tests passing.

Very similar to yours with some differences - I used `vector`s as containers and I have a `freelist` to reduce reallocations to some cases.

I do not have the last huge "Remove `NextAfter`" (https://github.com/solvespace/solvespace/pull/992/commits/3b395bb5a7cc09023c9cb0c9d14ea63a5f5e07bf) commit in the branch, but it is critical for some workloads.

Without this commit it is almost as fast (or barely slower) on this https://github.com/solvespace/solvespace/files/5962704/case-vents.slvs.gz from here https://github.com/solvespace/solvespace/issues/932#issue-806016900 compared to current master.

But it is drastically faster on setting the last translate to 20 copies in this https://github.com/solvespace/solvespace/files/5974674/LeliaManya2.zip. It takes about 2s while setting it to 6 copies takes 15s on master (I did not wait for 20 to finish)!

Next I will try yours on Windows with Visual Studio 2019.

**Comment by pjanx** [comment]
There should be a test for possible corner cases in `RemoveTagged`, I made a mistake there that passed but crashed on models.

I see how the freelist simplifies the algorithm in that method (at the cost of some more memory).

Like you, I also thought about putting some assertions in the iterator but... if they weren't there before... maybe it's not important.

I tried to change as little as possible but `NextAfter` just has to go.

Doing benchmarks was on my list, though the improvement was so immense that I just declared victory.

The next thing that irritates me on Solvespace is omnipresent NURBS failures and I can't really do math. How much do you with @phkahler charge? :) Triangle meshes also take a long time to compute.

**Comment by phkahler** [comment]
> The next thing that irritates me on Solvespace is omnipresent NURBS failures and I can't really do math. How much do you with @phkahler charge? :) Triangle meshes also take a long time to compute.

@pjanx  There is a NURBS tag for issues. The failures have been getting better and most are tracked in #738.  It's an area of the code that is a bit outside most peoples experience, but @ruevs and I have been chipping away at it for a while now. A few other people have had a look too which has been very helpful even when we didn't take their changes directly.

Triangle meshes are IMHO not to be used except when NURBS fail. The code there has not had the optimizations or OMP love applied so it could be slower now than doing the same with NURBS. When STL linking lands I suppose we'll hear more reports of mesh bugs.

SolveSpace usage has been increasing as have the number of contributors, so there is hope ;-)

**Comment by pjanx** [comment]
Yeah, I'm having some seemingly randomized mesh issues with my most complex model already (it NURBS-fails very early). Luckily Slicer mostly takes care of it on import, or the issues aren't too severe in 3D print.

**Comment by phkahler** [comment]
@pjanx the evil spirit patches replaces the list of things with a list of pointers to things. That made all the moving faster and probably reduced cache misses while searching.

**Comment by pjanx** [comment]
Replicating IDs in the index like he seems to be doing would eliminate some cache misses during search-by-ID, at the cost of even more memory. I'm not sure how much it is okay to waste on this--it'd be nice to have a memory profile. Binary search isn't particularly cache-friendly in any case, as it jumps around and prefetching fails to work optimally. In that sense, this general approach will always be slower during whole-list iteration. A lot of theorising here, besides the general observation of "large models are fairly snappy now".

**Comment by ruevs** [comment]
> There should be a test for possible corner cases in `RemoveTagged`, I made a mistake there that passed but crashed on models.

The test suite is far, far from complete/exhaustive/ achieving good code coverage.

I broke RemoveTagged as well until the last commit (the 0xDEADBEEF caught it ;-) )
`elemidx.resize(n);  // Clear left over elements at the end.`

To make it work I wrote a quick and dirty "unit test" of sorts for IdList with a simple (int, char) struct. But there is no good place to commit this.

**Comment by ruevs** [comment]
> How much do you with @phkahler charge? :)

For me it's just for fun ;-) Not to mention that I don't really feel qualified when hacking on the NURBS backend, I've forgotten all the linear algebra and calculus so long ago that I constantly have to consult references :-)

But having another mind to bounce ideas off of is extremely helpful. Thank you for that. For example you [starting to work on the IdList](https://github.com/solvespace/solvespace/pull/973) - it has been in the back of my mind since the [Whitequark commit and EvilSpirits patch](https://github.com/solvespace/solvespace/issues/939#issuecomment-810641820) (so at least four years), but you stirred things up, which lead to this pull request and my similar one.

Another one is Eigen - search the issues and pull requests ;-)

I'll make a tracking issue for performance - just as I did for NURBS in https://github.com/solvespace/solvespace/issues/738 .

**Comment by pjanx** [comment]
> This is one reason it would have been extremely hard to merge your last commit here when she was the maintainer. One reason SolveSpace is so stable is that she was (quite rightly) very careful about merging such all encompassing changes that are hard to regression test and also "dirty up" git blame, thus making regressions harder to track down. That said, I totally agree that `NextAfrer` absolutely has to go - it simply makes it impossible to make IdList efficient for all cases.

As far as my experience goes, 2.3 was much less stable, and a performance _minefield_ that made me re-learn to save often. Converting models to master is a pain, but the software is usable. I'm not sure whom to blame for this newfound stability. :)

I don't really like the "git blame" argument, it often seems like a universal excuse to never change anything. Bisection should take care of tracking regressions.

What I've noticed is that the codebase is in the process of losing some sense of coherence, which this PR doesn't particularly help. And cleaning that up is certainly an offensive against "git blame".

> I broke RemoveTagged as well until my last commit (the 0xDEADBEEF caught it ;-) )
> `elemidx.resize(n); // Clear left over elements at the end.`

I just read the source code one more time and found it. I've spent quite a few hours trying to visualise that function and put it to code, mostly because I said I would do it. Sadly I don't have any good methods for making things _simpler_.

**Comment by ruevs** [comment]
> Converting models to master is a pain, but the software is usable. I'm not sure whom to blame for this newfound stability. :)

[Whiteqark](https://github.com/solvespace/solvespace/graphs/contributors) ;-)

> I don't really like the "git blame" argument, it often seems like a universal excuse to never change anything. Bisection should take care of tracking regressions.

In general I agree, but contingent upon available maintainer time vs. desire for stability.

Anyway, I'm pretty sure either this/yours or mine, or a cleaned up amalgamation of the two (they are pretty much the same) will get merged after 3.0. 

( Both @phkahler and I have both been known to merge "stupid" stuff in the past ;-) )

File: src/dsc.h [hunk 1]
```diff
@@ -417,46 +441,38 @@ class IdList {
...
-        auto idx = std::distance(begin(), it);
-        auto i = static_cast<int>(idx);
-        return i;
+        return std::distance(begin(), LowerBound(t));
     }
+
     void ReserveMore(int howMuch) {
         if(n + howMuch > elemsAllocated) {
             elemsAllocated = n + howMuch;
+            int *newIndexes = (int *)::operator new[]((size_t)elemsAllocated*sizeof(int));
```
Comment (ruevs) [inline src/dsc.h:466]: [Using `vector` for the storage](https://github.com/ruevs/solvespace/blob/6bc1aa91c6a0f1acfc986b34ff54ed36efcdbfc1/src/dsc.h#L404) gets rid of the manual memory management, and hopefully at least in release mode should be as fast.

File: src/dsc.h [hunk 2]
```diff
@@ -467,16 +483,13 @@ class IdList {
         // Look to see if we already have something with the same handle value.
         ssassert(FindByIdNoOops(t->h) == nullptr, "Handle isn't unique");
 
-        // Find out where the added element should be.
+        // Shift indexes to the end of the array, if needed.
         int pos = LowerBoundIndex(*t);
+        memmove(indexes + pos + 1, indexes + pos, sizeof *indexes * (n - pos));
```
Comment (ruevs) [inline src/dsc.h:488]: I wonder if [insert in a `vector<int>`](https://github.com/ruevs/solvespace/blob/6bc1aa91c6a0f1acfc986b34ff54ed36efcdbfc1/src/dsc.h#L537) is as fast. I'll debug it tomorrow to see what it does.

File: src/dsc.h [hunk 3]
```diff
@@ -486,55 +499,21 @@ class IdList {
...
-    T &Get(size_t i) { return elem[i]; }
-    T const &Get(size_t i) const { return elem[i]; }
+    T &Get(size_t i) { return elem[indexes[i]]; }
+    T const &Get(size_t i) const { return elem[indexes[i]]; }
     T &operator[](size_t i) { return Get(i); }
     T const &operator[](size_t i) const { return Get(i); }
 
-    T *begin() { return IsEmpty() ? nullptr : &elem[0]; }
-    T *end() { return IsEmpty() ? nullptr : &elem[0] + n; }
-    const T *begin() const { return IsEmpty() ? nullptr : &elem[0]; }
```
Comment (ruevs) [inline src/dsc.h:534]: As above - the const iterators may be useful for compiler optimisations, if we can add `const` in some places where IdLists are used.

File: src/dsc.h [hunk 4]
```diff
@@ -548,23 +527,38 @@ class IdList {
...
-        for(src = 0; src < n; src++) {
-            if(elem[src].tag) {
-                // this item should be deleted
-                elem[src].Clear();
-            } else {
-                if(src != dest) {
-                    elem[dest] = elem[src];
-                }
-                dest++;
+        std::vector<int> elem2indexes(n);
```
Comment (ruevs) [inline src/dsc.h:530]: I need to trace through this with the debugger to really get it. Tomorrow.

**Review by ruevs — COMMENTED** [review]
OK so overall we did roughly the same. I'll steal your "remove NextAfter" commit and try mine with it tomorrow.

I think we should try to "converge" on a single solution. If you have time I would love to hear your comments/critique on my version.

File: src/dsc.h [hunk 5]
```diff
@@ -388,12 +388,36 @@ struct CompareId {
...
+        typedef T value_type;
+        typedef int difference_type;
+        typedef T *pointer;
+        typedef T &reference;
+
+        IdList<T, H> *list;
+        int offset;
+
+        Iterator(IdList<T, H> &list, int offset) : list(&list), offset(offset) {}
+        T &operator*() { return list->elem[list->indexes[offset]]; }
```
[the remark this answers is not in the document]
Comment (pjanx) [reply src/dsc.h:409]: That's one thing optimizers should be good at, unless some aliasing rules break it. I'd be somewhat surprised if it'd be measurable for the added complexity. Making `offset` point directly to the `indexes` array is more obvious. And both of these suffer if the array is resized during iteration, which STL just declares as undefined behaviour in such situations.

File: src/dsc.h [hunk 6]
```diff
@@ -388,12 +388,36 @@ struct CompareId {
...
+        T *operator->() { return &list->elem[list->indexes[offset]]; }
+        const T &operator*() const { return list->elem[list->indexes[offset]]; }
```
[the remark this answers is not in the document]
Comment (pjanx) [reply src/dsc.h:411]: MSVC required me to add `const` versions here, for `std::lower_bound`. See the first comment of the PR.

File: src/dsc.h [hunk 7]
```diff
@@ -388,12 +388,36 @@ struct CompareId {
...
+        const T *operator->() const { return &list->elem[list->indexes[offset]]; }
+        Iterator &operator--() { --offset; return *this; }
+        Iterator &operator++() { ++offset; return *this; }
+        Iterator &operator+=(int diff) { offset += diff; return *this; }
+        int operator-(const Iterator &i) const { return offset - i.offset; }
+        bool operator==(const Iterator &i) { return list == i.list && offset == i.offset; }
```
[the remark this answers is not in the document]
Comment (pjanx) [reply src/dsc.h:417]: It should rather be an assertion rather than a direct test like this.

File: src/dsc.h [hunk 8]
```diff
@@ -417,46 +441,38 @@ class IdList {
...
     }
 
-    T * LowerBound(H const& h) {
-        if(IsEmpty()) {
-            return nullptr;
-        }
-        auto it = std::lower_bound(begin(), end(), h, Compare());
-        return it;
+    Iterator LowerBound(H const& h) {
+        return std::lower_bound(begin(), end(), h, Compare());
```
[the remark this answers is not in the document]
Comment (pjanx) [reply src/dsc.h:456]: I wouldn't expect a meaningful difference, it boils down to a few pointer operations. Didn't you have to add a forward declaration and complicate `CompareId` because of that?

**Comment by ruevs** [comment]
> Converting models to master is a pain

By the way, why is this? Models from 2.0-2.3 should "just open™". If you have some that don't (and are not this https://github.com/solvespace/solvespace/issues/120) open an issue.

File: src/dsc.h (hunk 1 above)
Comment (pjanx) [reply src/dsc.h:466]: That's one of the things that should be unified throughout the codebase, I'm just using what's already there.

I don't even know how the custom allocator library is hooked up.

@Evil-Spirit wanted to use `realloc`, which should be a trick that `std::vector` could be able to do as well, though I'm dubious whether it's useful for geometric steps.

File: src/dsc.h (hunk 2 above)
Comment (pjanx) [reply src/dsc.h:488]: It should optimize that way.

**Comment by pjanx** [comment]
Mostly getting rid of all those extra degrees of freedom, rotation takes a while to figure out. I don't remember much else.

File: src/dsc.h (hunk 3 above)
Comment (pjanx) [reply src/dsc.h:534]: I'd be interested in seeing evidence of any difference in this case.

File: src/dsc.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (ruevs) [reply src/dsc.h:417]: It was an assertion before I removed it ;-)

File: src/dsc.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (ruevs) [reply src/dsc.h:456]: Yes. I called `CompareId" the predicate.

File: src/dsc.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (rpavlik) [reply src/dsc.h:411]: I'm inclined to think just implement the code that's needed on the YAGNI principle. Unused methods are untested and possible traps for the future.

File: src/dsc.h (hunk 1 above)
Comment (rpavlik) [reply src/dsc.h:466]: The important bit is that idlist is reference semantics, not value semantics, so you can't embed a std::vector, but rather a std::shared_ptr<std::vector> - a lesson I learned painfully.

**Comment by ruevs** [comment]
@pjanx 
Tried your version - it seems to work just as fast and as correctly as mine. RemoveTagged seems correct, but it's a bit complicated without a free element list.

I created a pull request with my version with a cleaned up history: https://github.com/solvespace/solvespace/pull/1002
It uses your "remove NextAfter" commit.

Next I will try to see if I can measure speed differences between our implementations but I highly doubt it that there will be any. 
I will also profile to see if there are some more bottlenecks related to IdList.

After that it boils down to style/aesthetics etc.

**Comment by pjanx** [comment]
I think we need an arbiter. It'd be nice to have a memory profile as well, Valgrind can also do that, IIRC.

**Comment by Evil-Spirit** [comment]
Hi there!

Things I remeber:

1.  memove and realloc is superfast
2.  if you are doing containers yourserf it is a bad idea to use another containers like std::vector. you can't control things under the hood of this. If you afraid of manual memory management - just remember, you are doing container for not doing manual memory management around whole program, but inside - you have to do it - instead, you will lost all the benefits of doing this job
3.  when you need map-lookup, you should use binary search. when you need interaing over elements - you may iterate only linear array of data. But NOT ALL THE TIME. Some operations should be performed on sorted arrays, like saving file. You can just break round-trip tests by not iterating in the hv alpahbetic order.

File: src/dsc.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (ruevs) [reply src/dsc.h:411]: Removed all unused methods from mine.

File: src/dsc.h (hunk 1 above)
Comment (ruevs) [reply src/dsc.h:466]: @rpavlik Do you mean to say that once a pointer to an element is "taken out' of an IdList it may be kept around and dereferenced after the list has been modified? And thus the pointer may be invalid?

If this is the case can you point to an example?

But this does not make sense - the original implementation of IdList did not guarantee that the elemnts will not move - in fact it moved them around all the time!

In fact my implementation is "better" in this respect since it never moves elements once they are added because the `elemstore` is not rearranged when elements are removed:
https://github.com/ruevs/solvespace/blob/3c6e2be4b920d706d8231030467c5271a6d87d85/src/dsc.h#L570
the removed elements are simply added to the `freelist`.

**Comment by baryluk** [comment]
A bit of my input. With concerns from #932 

I didn't check the memory usage, but did a bit of speed testing. For changing constraints and these PRs. Just dragging unconstrainted extrusion length by mouse, and looking when the UI is fully interactive again. I used gcc 10.2.1 on Linux, amd64. `-O3 -flto -march=native`. With no optimisations, these PRs are actually slower than current master, because of extra small functions and indirections in the iterators. But with optimizations on, they are a tiny bit faster. 

master, -O3 -flto: ~6.6s, 2.78s for Generate::DIRTY
ruevs, -O3 -flto: ~5.4s, 2.31s for Generate::DIRTY
pjanx, -O3 -flto: ~5.5s, 2.27s for Generate::DIRTY

That doesn't sounds like a lot to me, and I don't think it is worth merging any of them it at the current state. Don't get me wrong, these is good code, and a move in a right direction, but it might require more profiling, tuning and base data structure selection.

(No OpenMP used).

**Comment by ruevs** [comment]
> A bit of my input. With concerns from #932
> 
> I didn't check the memory usage, but did a bit of speed testing. For changing constraints and these PRs. Just dragging unconstrainted extrusion length by mouse, and looking when the UI is fully interactive again. I used gcc 10.2.1 on Linux, amd64. `-O3 -flto -march=native`. With no optimisations, these PRs are actually slower than current master, because of extra small functions and indirections in the iterators. But with optimizations on, they are a tiny bit faster.
> 
> master, -O3 -flto: ~6.6s, 2.78s for Generate::DIRTY
> ruevs, -O3 -flto: ~5.4s, 2.31s for Generate::DIRTY
> pjanx, -O3 -flto: ~5.5s, 2.27s for Generate::DIRTY
> 
> That doesn't sounds like a lot to me, and I don't think it is worth merging any of them it at the current state. Don't get me wrong, these is good code, and a move in a right direction, but it might require more profiling, tuning and base data structure selection.
> 
> (No OpenMP used).
> 
> UPDATE:
> 
> LeilaManya2, changing last translate group from 3 to 20 repeats.
> 
> master: ~138s, 126119ms for Generate::DIRTY
> pjanx: ~1.2s, 427ms for Generate::DIRTY
> ruevs: ~1.1s, 403ms for Generate::DIRTY
> 
> Nice! Ok. I might be convinced.

I was just about to mention https://github.com/solvespace/solvespace/files/5974674/LeliaManya2.zip when I noticed you updated your comment ;-) Thanks for running it to 20 repeats on master - I did not have the patience :-)

Would you please do the same timing comparison when modifying some dimension in the first group of this model? https://github.com/solvespace/solvespace/issues/841#issuecomment-742153838

How do you do the measurements on Linux?

**Comment by phkahler** [comment]
> By the way, why is this? Models from 2.0-2.3 should "just open™". If you have some that don't (and are not this #120) open an issue.

Off topic, but I know some Lathe groups can have issues. Lathe used to create 2 copies of the sketch entities like at the start and end of an extrude or revolve, even though both copies were exactly the same. I removed those redundant entities so any sketch that constrained against them could break in 3.0 depending which copy the constraint used.

**Comment by phkahler** [comment]
Going with the @ruevs version unless something bad happens.
